### PR TITLE
Return resources in the API v2 simple ontology schema

### LINKED ISSUE #725: Allow resources to be returned in the simple schema
Currently only the complex schema can be returned.

### LINKED ISSUE #797: GenerateContributorsFile broken
@benjamingeer could you please take a look when you have time? The following error happens with and without a supplied token:

```
sbt:webapi> runMain org.knora.webapi.util.GenerateContributorsFile -t myprivatetoken
[info] Packaging /Users/subotic/_github.com/knora/webapi/target/scala-2.12/webapi_2.12-1.3.0.jar ...
[info] Done packaging.
[info] Running (fork) org.knora.webapi.util.GenerateContributorsFile -t bcb20132ec1ed3bbc2fbd674792a4b81aa4f4bb2
[error] objc[3238]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk/Contents/Home/jre/bin/java (0x1038b84c0) and /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk/Contents/Home/jre/lib/libinstrument.dylib (0x1039504e0). One of the two will be used. Which one is undefined.
[error] Exception in thread "main" java.lang.ClassCastException: spray.json.JsNull$ cannot be cast to spray.json.JsString
[error] 	at org.knora.webapi.util.GenerateContributorsFile$.$anonfun$contributorsWithNames$1(GenerateContributorsFile.scala:71)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:929)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:929)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
[error] 	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
[error] 	at org.knora.webapi.util.GenerateContributorsFile$.delayedEndpoint$org$knora$webapi$util$GenerateContributorsFile$1(GenerateContributorsFile.scala:66)
[error] 	at org.knora.webapi.util.GenerateContributorsFile$delayedInit$body.apply(GenerateContributorsFile.scala:35)
[error] 	at scala.Function0.apply$mcV$sp(Function0.scala:34)
[error] 	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
[error] 	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
[error] 	at scala.App.$anonfun$main$1$adapted(App.scala:76)
[error] 	at scala.collection.immutable.List.foreach(List.scala:389)
[error] 	at scala.App.main(App.scala:76)
[error] 	at scala.App.main$(App.scala:74)
[error] 	at org.knora.webapi.util.GenerateContributorsFile$.main(GenerateContributorsFile.scala:35)
[error] 	at org.knora.webapi.util.GenerateContributorsFile.main(GenerateContributorsFile.scala)
[error] java.lang.RuntimeException: Nonzero exit code returned from runner: 1
[error] 	at sbt.ForkRun.processExitCode$1(Run.scala:33)
[error] 	at sbt.ForkRun.run(Run.scala:42)
[error] 	at sbt.Defaults$.$anonfun$bgRunMainTask$6(Defaults.scala:1147)
[error] 	at sbt.Defaults$.$anonfun$bgRunMainTask$6$adapted(Defaults.scala:1142)
[error] 	at sbt.internal.BackgroundThreadPool.$anonfun$run$1(DefaultBackgroundJobService.scala:366)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at scala.util.Try$.apply(Try.scala:209)
[error] 	at sbt.internal.BackgroundThreadPool$BackgroundRunnable.run(DefaultBackgroundJobService.scala:289)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
[error] (Compile / runMain) Nonzero exit code returned from runner: 1
[error] Total time: 7 s, completed Mar 28, 2018 10:15:02 AM
```

### LINKED ISSUE #835: Incorrect representation of IRI object values in JSON-LD
It had been bothering me for a while that there seemed to be no type information attached to IRIs that are object values in JSON-LD. We've been representing them like this:

```jsonld
{
  "knora-api:subjectType" : "http://api.knora.org/ontology/knora-api/simple/v2#Region"
}
```

This is a problem for both input and output: for input, because we have to parse the string object to guess if it's supposed to be an IRI, and for output, because for #609 we need to convert JSON-LD to Turtle and format the IRIs correctly.

It also bothered me that there seemed to be no way to get the JSON-LD API to use ontology prefix labels in these IRIs.

I checked the JSON-LD spec, and it turns out we've been doing this wrong. [Object values that are IRIs are supposed to look like this](https://json-ld.org/spec/latest/json-ld/#iris):

```jsonld
"knora-api:subjectType" : {
  "@id" : "knora-api:Region"
}
```

### PR DESCRIPTION
**Pull request by benjamingeer**
- [x] Refactor `ReadResourcesSequenceV2` and its components, separating schema conversion from JSON-LD generation.
- [x] Enable `ReadResourcesSequenceV2` and its components to return JSON-LD in the simple schema.
- [x] Use `rdfs:label` in responses instead of `schema:name` **(requires a change in SALSAH 2)**.
- [x] Enable `ResourcesRouteV2` to handle requests that ask for a response in either the simple or the complex schema, using an HTTP header (`X-Knora-Accept-Schema`) or a URL parameter (`schema`).
- [x] Fix the format of the string representation of dates in both schemas, so it conforms to `StringFormatter.KnoraDateRegex`.
- [x] Represent IRI values correctly in JSON-LD (#835), instead of as simple strings **(requires a change in SALSAH 2)**.
- [x] Add tests.
- [x] Update docs.
- [x] Update release notes.

Resolves #725.
Fixes #835.
Fixes #797.

**Requires https://github.com/dhlab-basel/Salsah/pull/245 to be merged**

### CODE REVIEW

**Comment by benjamingeer** [comment]
@lrosenth I've been thinking about how we can handle content negotiation when using ARK URLs for resources. I suggest it could work like this:

If the client is a browser, they'll probably send an HTTP `Accept` header like this when they open the ARK URL:

```
Accept: text/html
```

Then our ARK server can redirect them to the SALSAH GUI.

If the client wants JSON or JSON-LD, they'll probably send an `Accept` header like this:

```
Accept: application/json,application/ld+json
```

In that case, we could assume that they don't know much about Knora (otherwise they wouldn't be using the ARK URL), and that they want something like Linked Open Data, i.e. something simple and interoperable. So ARK server can redirect them directly to the Knora API to get the resource, and add this header:

```
X-Knora-Accept-Schema: simple
```

Then they'll get the resource as JSON-LD, using the simple schema.

How does that sound?

**Comment by benjamingeer** [comment]
...or instead of using `X-Knora-Accept-Schema`, the  ARK server could just add `?schema=simple` to the Knora API URL.

**Comment by subotic** [comment]
If I may. Sounds good to me. Are we going to support both `X-Knora-Accept-Schema` and `?schema=simple`?

**Comment by benjamingeer** [comment]
> Sounds good to me. Are we going to support both X-Knora-Accept-Schema and ?schema=simple?

Both are supported in this PR.

File: Contributors.md [hunk 1]
```diff
@@ -11,5 +11,6 @@
     - Ivan Subotic <https://github.com/subotic>
     - Loïc Jaouen <https://github.com/loicjaouen>
     - Sepideh Alassi <https://github.com/SepidehAlassi>
-    - André Kilchenmann <https://github.com/kilchenmann>
     - Stefan Münnich <https://github.com/musicEnfanthen>
+    - André Kilchenmann <https://github.com/kilchenmann>
+    - <https://github.com/janCstoffregen>
```
Comment (tobiasschweizer) [inline Contributors.md]: Shouldn't that be "Jan Stoffregen <https://github.com/janCstoffregen>"`

Comment (benjamingeer) [reply Contributors.md]: It can't be, because he hasn't put his name in his GitHub accont.

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/OntologyV2R2RSpec.scala [hunk 2]
```diff
@@ -270,46 +266,16 @@ class OntologyV2R2RSpec extends R2RSpec {
             }
         }
 
-        "serve a project-specific ontology whose IRI contains a project ID, as JSON-LD, using the simple schema" in {
```
Comment (tobiasschweizer) [inline webapi/src/test/scala/org/knora/webapi/e2e/v2/OntologyV2R2RSpec.scala]: Why did you remove these test?

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala [hunk 3]
```diff
@@ -93,6 +94,32 @@ class ResourcesRouteV2R2Spec extends R2RSpec {
             }
         }
 
+        "perform a resource request for the book 'Reise ins Heilige Land' using the simple schema (specified by an HTTP header)" in {
+
+            Get(s"/v2/resources/${URLEncoder.encode("http://rdfh.ch/2a6221216701", "UTF-8")}").addHeader(new SchemaHeader(RouteUtilV2.SIMPLE_SCHEMA_NAME)) ~> resourcesPath ~> check {
```
Comment (tobiasschweizer) [inline webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala]: `addHeader(new SchemaHeader(RouteUtilV2.SIMPLE_SCHEMA_NAME)`

Wouldn't it be better for a test to send the complete header as a string: "x-knora-accept-schema: simple"

In order to make sure that the parsing works correctly.

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ResponseCheckerR2RV2.scala [hunk 4]
```diff
@@ -209,13 +177,10 @@ object ResponseCheckerR2RV2 {
...
+        val expectedJsonLDDocument = JsonLDUtil.parseJsonLD(expectedJSONLD)
+        val receivedJsonLDDocument = JsonLDUtil.parseJsonLD(receivedJSONLD)
 
-        val expectedResponseAsScala: Map[IRI, Any] = JSONToScala(expectedJSONLD)
-
-        val receivedResponseAsScala: Map[IRI, Any] = JSONToScala(receivedJSONLD)
-
-        assert(expectedResponseAsScala == receivedResponseAsScala, "Mapping creation request response did not match expected response")
-
+        assert(expectedJsonLDDocument == receivedJsonLDDocument, "Mapping creation request response did not match expected response")
```
Comment (tobiasschweizer) [inline webapi/src/test/scala/org/knora/webapi/e2e/v2/ResponseCheckerR2RV2.scala]: Should we add a test to make sure this method works correctly?

We could add a test the expects this method to fail. I am thinking of the `received == received` or `expected == expcted` bug ...

Comment (tobiasschweizer) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ResponseCheckerR2RV2.scala]: I know there is no logical way to rule out errors. So I won't ask for a test that checks the tests for the test ...

File: docs/src/paradox/03-webapi/api-v2/ontology-information.md [hunk 5]
```diff
@@ -47,27 +47,45 @@ To get metadata about all ontologies:
 HTTP GET to http://host/v2/ontologies/metadata
 ```
 
-The response is in the default API v2 schema. Sample response:
+The response is in the complex API v2 schema. Sample response:
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply docs/src/paradox/03-webapi/api-v2/ontology-information.md]: Can I do this in a later pull request?

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala (hunk 3 above)
Comment (benjamingeer) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala]: There is no way to do that with Akka-HTTP. A custom header has to be specified as a class. There's a link to the relevant docs in `SchemaHeader.scala`.

Comment (tobiasschweizer) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala]: I think this is because the tests are not actual HTTP requests as the browser would send them.

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/ontologymessages/OntologyMessagesV2.scala [hunk 6]
```diff
@@ -1921,25 +1915,55 @@ object EntityInfoContentV2 {
...
+                            PredicateInfoV2(
+                                predicateIri = predicateIri,
+                                objects = objArray.value.map {
+                                    case jsonObjElem: JsonLDObject => SmartIriLiteralV2(JsonLDUtil.iriFromJsonLDObject(jsonObjElem, stringFormatter.toSmartIriWithErr))
+                                    case other => throw AssertionException(s"Invalid object for predicate $predicateIriStr: $other")
+                                }
+                            )
+                        } else if (objArray.value.forall {
+                            case jsonObjElem: JsonLDObject if JsonLDUtil.isStringWithLang(jsonObjElem) =>
+                                // All the elements of the array are strings with language codes.
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/ontologymessages/OntologyMessagesV2.scala]: When an ontology entity is created or modified.

[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/ontologymessages/OntologyMessagesV2.scala]: Ok, got it.

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala (hunk 3 above)
Comment (benjamingeer) [reply webapi/src/test/scala/org/knora/webapi/e2e/v2/ResourcesRouteV2R2Spec.scala]: This isn't anything specific to testing, it seems to be inherent in the design of the Akka HTTP client API:

https://doc.akka.io/docs/akka-http/current/common/http-model.html#custom-headers

I didn't find any other way to add a custom header to an HTTP request in Akka.
